src/System.Runtime/tests/System/DateTimeTests.cs

```diff
@@ -2177,6 +2178,59 @@ public void GetObjectData_NullInfo_ThrowsArgumentNullException()
             AssertExtensions.Throws<ArgumentNullException>("info", () => ((ISerializable)DateTime.Now).GetObjectData(null, new StreamingContext()));
         }
 
+        [Fact]
+        public void TestRoundTrippingDateTimeAndFileTime()
+        {
+            // This test ensure the round tripping of DateTime with the system file time.
+            // It is important to have this working on systems support leap seconds as the conversion wouldn't be simple
+            // conversion but involve some OS calls to ensure the right conversion is happening.
+
+            DateTime now = DateTime.UtcNow;
+            long fileTime = now.ToFileTimeUtc();
+            DateTime roundTrippedDateTime = DateTime.FromFileTimeUtc(fileTime);
+            Assert.Equal(now, roundTrippedDateTime);
+
+            now = DateTime.Now;
+            fileTime = now.ToFileTime();
+            roundTrippedDateTime = DateTime.FromFileTime(fileTime);
+            Assert.Equal(now, roundTrippedDateTime);
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void TestTimeSynchronizationWithTheSystem()
+        {
+            // The reported time by the framework should be synchronized with the OS.
+            // There shouldn't be any shift by more than one second, otherwise there is something wrong.
+            // This test is useful when running on a system supporting leap seconds to ensure when the system
+            // has leap seconds, the framework reported time will still be synchronized.
+
+            SYSTEMTIME st;
+            GetSystemTime(out st);
+            DateTime dt = DateTime.UtcNow;
+
+            DateTime systemDateTimeNow  = new DateTime(st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond, st.wMillisecond, DateTimeKind.Utc);
+            TimeSpan diff = systemDateTimeNow - dt;
+
+            Assert.True(diff < TimeSpan.FromSeconds(1), $"Reported DateTime.UtcNow {dt} is shifted by more than one second then the system time {systemDateTimeNow}");
+        }
+
+        [DllImport("Kernel32.dll")]
+        internal static extern void GetSystemTime(out SYSTEMTIME lpSystemTime);
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct SYSTEMTIME
+        {
+            internal ushort wYear;
+            internal ushort wMonth;
+            internal ushort wDayOfWeek;
+            internal ushort wDay;
+            internal ushort wHour;
+            internal ushort wMinute;
+            internal ushort wSecond;
+            internal ushort wMillisecond;
+        }
+
```

Comment:
Do we have or need tests that constructs a specific leap second time and validates that the right things happen?

Reply:
I don't think we can do that in the current time as we don't have a system with a leap seconds.